### ISSUE TITLE
Fallback ACS URL gets reused across requests under Devise

### ISSUE THREAD

**Issue by md5** [PracticallyGreen/omniauth-saml#54]
I just spent a bunch of time debugging an issue using `omniauth-saml` with Devise. We are using the "fallback" functionality introduced in #15 to dynamically set our `:assertion_consumer_service_url` to the Omniauth `callback_url`.

The issue we ran into was that the `:assertion_consumer_service_url` was being calculated and cached based on the `callback_url` of the first request to be handled. This was then being used for subsequent requests, even if the hostname for the request was not the same, resulting in the user being sent to the wrong hostname after logging in with the IdP.

I wasn't able to determine whether this is a bug in `rails`, `devise`, `omniauth`, or `omniauth-saml`, but I was able to work around it by monkey-patching this strategy like so:

```ruby
class OmniAuth::Strategies::SAML
  alias_method :original_request_phase, :request_phase
  def request_phase
    options[:assertion_consumer_service_url] = nil
    original_request_phase
  end
end
```

During my debugging, I was able to determine that a different instance of `OmniAuth::Strategies::SAML` was being used for each call to `request_phase`, but that those instances all shared the same instance of `OmniAuth::Strategy::Options`. However, when I added an `initialize` method to my monkey-patched version of `OmniAuth::Strategies::SAML`, I did not see it getting called more than once.

I'm still going to try to reproduce this in a minimal test case, but I wanted to open this issue in case anyone else has observed this problem.

A simple fix in the case of `omniauth-saml` might be to avoid any code that mutates `options`, replacing it with code that makes a copy of `options` first before making any changes it needs (e.g. the defaulting of `options[:assertion_consumer_service_url]`.

**Comment by md5** [comment]
BTW, I should add that my understanding is that a new instance of `OmniAuth::Strategies::SAML` should be created for each HTTP request.

I believe the instance itself should be created by the `ActionDispatch::MiddlewareStack::Middleware` class here: https://github.com/rails/rails/blob/4-2-stable/actionpack/lib/action_dispatch/middleware/stack.rb#L43

I would then expect it to call the `initialize` method of the middleware class in question via `new`. This should result in it creating a new copy of `default_options` in ``OmniAuth::Strategy`, which is here: https://github.com/intridea/omniauth/blob/c3e3aa5bbfb589db1ebed40324c7eea2e3153f8c/lib/omniauth/strategy.rb#L132

**Comment by md5** [comment]
Turns out my understanding of the middleware lifecycle was incorrect. In general, the Rails middleware stack is created once and the same middleware instances are used throughout the application lifecycle. When I made my own middleware and checked the `object_id` on each `call`, it was not changing.

It turns out that OmniAuth was throwing me off with the changing `object_id` thing because it does a `dup` on each `call()`. See here: https://github.com/intridea/omniauth/blob/c3e3aa5bbfb589db1ebed40324c7eea2e3153f8c/lib/omniauth/strategy.rb#L164

I can see two fixes here. One would be to propose that `OmniAuth::Strategy` have an implementation of `initialize_copy` that adds a call to `@options = @options.clone`. The other would be for `omniauth-saml` specifically to do its own `saml_options = options.clone` before making any mutating changes to the options it passes to `ruby-saml`.

I'm going to make a PR for the latter approach against this repo. I think I'll also open an issue for OmniAuth to discuss whether `@options` should be `dup`'d when the strategy itself gets `dup`'d.

**Comment by md5** [comment]
Just raised an issue in the OmniAuth repo: https://github.com/intridea/omniauth/issues/823

**Comment by md5** [comment]
I'm closing this since I think the underlying issue is clear and is covered either by #56 or intridea/omniauth#823

**Comment by md5** [comment]
See https://github.com/PracticallyGreen/omniauth-saml/pull/63 for the latest proposed fix since the fix in #56 was Rails-specific.
